# 项目：用线性回归预测房价数据

## 分析目标

此数据分析报告的目的是，基于已有的房屋销售价格，以及有关该房屋的属性，进行线性回归分析，从而利用得到的线性回归模型，能对以下未知售价的房屋根据属性进行价格预测：

面积为6500平方英尺，有4个卧室、2个厕所，总共2层，不位于主路，无客人房，带地下室，有热水器，没有空调，车位数为2，位于城市首选社区，简装修。

## 简介

数据集`house_price.csv`记录了超过五百栋房屋的交易价格，以及房屋的相关属性信息，包括房屋面积、卧室数、厕所数、楼层数、是否位于主路、是否有客房，等等。

`house_price.csv`每列的含义如下：
- price：房屋出售价格
- area：房屋面积，以平方英尺为单位
- bedrooms：卧室数
- bathrooms：厕所数
- stories：楼层数
- mainroad：是否位于主路
   - yes  是
   - no	  否
- guestroom：是否有客房
   - yes  是
   - no	  否
- basement：是否有地下室
   - yes  是
   - no	  否
- hotwaterheating：是否有热水器
   - yes  是
   - no	  否
- airconditioning：是否有空调
   - yes  是
   - no	  否
- parking：车库容量，以车辆数量为单位
- prefarea：是否位于城市首选社区
   - yes  是
   - no	  否
- furnishingstatus：装修状态
   - furnished       精装
   - semi-furnished	 简装
   - unfurnished     毛坯

# 读取数据

In [1]:
import statsmodels.api as sm
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
original_house_price=pd.read_csv(r"C:\Users\YU\Downloads\house_price.csv")
original_house_price.head()

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

# 评估和清理数据

In [3]:
cleaned_house_price=original_house_price.copy()

## 数据整齐度

In [4]:
cleaned_house_price.head(10)

price   area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000   7420         4          2        3      yes        no       no   
1  12250000   8960         4          4        4      yes        no       no   
2  12250000   9960         3          2        2      yes        no      yes   
3  12215000   7500         4          2        2      yes        no      yes   
4  11410000   7420         4          1        2      yes       yes      yes   
5  10850000   7500         3          3        1      yes        no      yes   
6  10150000   8580         4          3        4      yes        no       no   
7  10150000  16200         5          3        2      yes        no       no   
8   9870000   8100         4          1        2      yes       yes      yes   
9   9800000   5750         3          2        4      yes       yes       no   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished  
5              no             yes        2      yes   semi-furnished  
6              no             yes        2      yes   semi-furnished  
7              no              no        0       no      unfurnished  
8              no             yes        2      yes        furnished  
9              no             yes        1      yes      unfurnished

## 数据干净度

In [5]:
cleaned_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [6]:
cleaned_house_price['mainroad']=cleaned_house_price['mainroad'].astype('category')
cleaned_house_price['guestroom']=cleaned_house_price['guestroom'].astype('category')
cleaned_house_price['basement']=cleaned_house_price['basement'].astype('category')
cleaned_house_price['hotwaterheating']=cleaned_house_price['hotwaterheating'].astype('category')
cleaned_house_price['airconditioning']=cleaned_house_price['airconditioning'].astype('category')
cleaned_house_price['prefarea']=cleaned_house_price['prefarea'].astype('category')
cleaned_house_price['furnishingstatus']=cleaned_house_price['furnishingstatus'].astype('category')

In [7]:
cleaned_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int64   
 1   area              545 non-null    int64   
 2   bedrooms          545 non-null    int64   
 3   bathrooms         545 non-null    int64   
 4   stories           545 non-null    int64   
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int64   
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int64(6)
memory usage: 30.3 KB


### 处理缺失数据

从info方法的输出结果来看，cleaned_house_price不存在缺失值，因此不需要对缺失数据进行处理。

### 处理重复数据

根据数据变量的含义以及内容来看，允许变量重复，我们不需要对此数据检查是否存在重复值。

### 处理不一致数据

In [8]:
cleaned_house_price['mainroad'].value_counts()

mainroad
yes    468
no      77
Name: count, dtype: int64

In [9]:
cleaned_house_price['guestroom'].value_counts()

guestroom
no     448
yes     97
Name: count, dtype: int64

In [10]:
cleaned_house_price['basement'].value_counts()

basement
no     354
yes    191
Name: count, dtype: int64

In [11]:
cleaned_house_price['hotwaterheating'].value_counts()

hotwaterheating
no     520
yes     25
Name: count, dtype: int64

In [12]:
cleaned_house_price['airconditioning'].value_counts()

airconditioning
no     373
yes    172
Name: count, dtype: int64

In [13]:
cleaned_house_price['prefarea'].value_counts()

prefarea
no     417
yes    128
Name: count, dtype: int64

In [14]:
cleaned_house_price['furnishingstatus'].value_counts()

furnishingstatus
semi-furnished    227
unfurnished       178
furnished         140
Name: count, dtype: int64

从以上输出结果来看，均不存在不一致数据。

### 处理无效或错误数据

In [15]:
cleaned_house_price.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

从以上统计信息来看，cleaned_house_price里不存在脱离现实意义的数值。

# 分析数据

In [16]:
lr_house_price = cleaned_house_price.copy()

In [18]:
lr_house_price=pd.get_dummies(lr_house_price,columns=['mainroad','guestroom',
                                                      'basement','hotwaterheating',
                                                      'airconditioning','prefarea',
                                                      'furnishingstatus'],dtype=int,drop_first=True)
lr_house_price

price  area  bedrooms  bathrooms  stories  parking  mainroad_yes  \
0    13300000  7420         4          2        3        2             1   
1    12250000  8960         4          4        4        3             1   
2    12250000  9960         3          2        2        2             1   
3    12215000  7500         4          2        2        3             1   
4    11410000  7420         4          1        2        2             1   
..        ...   ...       ...        ...      ...      ...           ...   
540   1820000  3000         2          1        1        2             1   
541   1767150  2400         3          1        1        0             0   
542   1750000  3620         2          1        1        0             1   
543   1750000  2910         3          1        1        0             0   
544   1750000  3850         3          1        2        0             1   

     guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
0                0             0                    0                    1   
1                0             0                    0                    1   
2                0             1                    0                    0   
3                0             1                    0                    1   
4                1             1                    0                    1   
..             ...           ...                  ...                  ...   
540              0             1                    0                    0   
541              0             0                    0                    0   
542              0             0                    0                    0   
543              0             0                    0                    0   
544              0             0                    0                    0   

     prefarea_yes  furnishingstatus_semi-furnished  \
0               1                                0   
1               0                                0   
2               1                                1   
3               1                                0   
4               0                                0   
..            ...                              ...   
540             0                                0   
541             0                                1   
542             0                                0   
543             0                                0   
544             0                                0   

     furnishingstatus_unfurnished  
0                               0  
1                               0  
2                               0  
3                               0  
4                               0  
..                            ...  
540                             1  
541                             0  
542                             1  
543                             0  
544                             1  

[545 rows x 14 columns]

In [19]:
y=lr_house_price['price']
x=lr_house_price.drop('price',axis=1)

一般我们认为，当相关系数的绝对值大于0.8的时候，可能导致严重共线性，所以我们检查的时候，找绝对值大于0.8的值即可。

In [20]:
x.corr().abs()>0.8

area  bedrooms  bathrooms  stories  parking  \
area                              True     False      False    False    False   
bedrooms                         False      True      False    False    False   
bathrooms                        False     False       True    False    False   
stories                          False     False      False     True    False   
parking                          False     False      False    False     True   
mainroad_yes                     False     False      False    False    False   
guestroom_yes                    False     False      False    False    False   
basement_yes                     False     False      False    False    False   
hotwaterheating_yes              False     False      False    False    False   
airconditioning_yes              False     False      False    False    False   
prefarea_yes                     False     False      False    False    False   
furnishingstatus_semi-furnished  False     False      False    False    False   
furnishingstatus_unfurnished     False     False      False    False    False   

                                 mainroad_yes  guestroom_yes  basement_yes  \
area                                    False          False         False   
bedrooms                                False          False         False   
bathrooms                               False          False         False   
stories                                 False          False         False   
parking                                 False          False         False   
mainroad_yes                             True          False         False   
guestroom_yes                           False           True         False   
basement_yes                            False          False          True   
hotwaterheating_yes                     False          False         False   
airconditioning_yes                     False          False         False   
prefarea_yes                            False          False         False   
furnishingstatus_semi-furnished         False          False         False   
furnishingstatus_unfurnished            False          False         False   

                                 hotwaterheating_yes  airconditioning_yes  \
area                                           False                False   
bedrooms                                       False                False   
bathrooms                                      False                False   
stories                                        False                False   
parking                                        False                False   
mainroad_yes                                   False                False   
guestroom_yes                                  False                False   
basement_yes                                   False                False   
hotwaterheating_yes                             True                False   
airconditioning_yes                            False                 True   
prefarea_yes                                   False                False   
furnishingstatus_semi-furnished                False                False   
furnishingstatus_unfurnished                   False                False   

                                 prefarea_yes  \
area                                    False   
bedrooms                                False   
bathrooms                               False   
stories                                 False   
parking                                 False   
mainroad_yes                            False   
guestroom_yes                           False   
basement_yes                            False   
hotwaterheating_yes                     False   
airconditioning_yes                     False   
prefarea_yes                             True   
furnishingstatus_semi-furnished         False   
furnishingstatus_unfurnished            False   

                                 furnish

从以上输出来看，不同变量之间相关系数的绝对值均小于或等于0.8。

接下来，给模型的线性方程添加截距。

In [21]:
x=sm.add_constant(x)
x

const  area  bedrooms  bathrooms  stories  parking  mainroad_yes  \
0      1.0  7420         4          2        3        2             1   
1      1.0  8960         4          4        4        3             1   
2      1.0  9960         3          2        2        2             1   
3      1.0  7500         4          2        2        3             1   
4      1.0  7420         4          1        2        2             1   
..     ...   ...       ...        ...      ...      ...           ...   
540    1.0  3000         2          1        1        2             1   
541    1.0  2400         3          1        1        0             0   
542    1.0  3620         2          1        1        0             1   
543    1.0  2910         3          1        1        0             0   
544    1.0  3850         3          1        2        0             1   

     guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
0                0             0                    0                    1   
1                0             0                    0                    1   
2                0             1                    0                    0   
3                0             1                    0                    1   
4                1             1                    0                    1   
..             ...           ...                  ...                  ...   
540              0             1                    0                    0   
541              0             0                    0                    0   
542              0             0                    0                    0   
543              0             0                    0                    0   
544              0             0                    0                    0   

     prefarea_yes  furnishingstatus_semi-furnished  \
0               1                                0   
1               0                                0   
2               1                                1   
3               1                                0   
4               0                                0   
..            ...                              ...   
540             0                                0   
541             0                                1   
542             0                                0   
543             0                                0   
544             0                                0   

     furnishingstatus_unfurnished  
0                               0  
1                               0  
2                               0  
3                               0  
4                               0  
..                            ...  
540                             1  
541                             0  
542                             1  
543                             0  
544                             1  

[545 rows x 14 columns]

下一步就可以调用OLS函数，利用最小二乘法来得到线性回归模型的参数值。

In [22]:
model=sm.OLS(y,x).fit()

In [23]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.682
Model:                            OLS   Adj. R-squared:                  0.674
Method:                 Least Squares   F-statistic:                     87.52
Date:                Wed, 20 Nov 2024   Prob (F-statistic):          9.07e-123
Time:                        15:43:10   Log-Likelihood:                -8331.5
No. Observations:                 545   AIC:                         1.669e+04
Df Residuals:                     531   BIC:                         1.675e+04
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
const                            4.277e+04   2.64e+05      0.162      0.872   -4.76e+05    5.62e+05
area                              244.1394     24.289     10.052      0.000     196.425     291.853
bedrooms                         1.148e+05   7.26e+04      1.581      0.114   -2.78e+04    2.57e+05
bathrooms                        9.877e+05   1.03e+05      9.555      0.000    7.85e+05    1.19e+06
stories                          4.508e+05   6.42e+04      7.026      0.000    3.25e+05    5.77e+05
parking                          2.771e+05   5.85e+04      4.735      0.000    1.62e+05    3.92e+05
mainroad_yes                     4.213e+05   1.42e+05      2.962      0.003    1.42e+05    7.01e+05
guestroom_yes                    3.005e+05   1.32e+05      2.282      0.023    4.18e+04    5.59e+05
basement_yes                     3.501e+05    1.1e+05      3.175      0.002    1.33e+05    5.67e+05
hotwaterheating_yes              8.554e+05   2.23e+05      3.833      0.000    4.17e+05    1.29e+06
airconditioning_yes               8.65e+05   1.08e+05      7.983      0.000    6.52e+05    1.08e+06
prefarea_yes                     6.515e+05   1.16e+05      5.632      0.000    4.24e+05    8.79e+05
furnishingstatus_semi-furnished -4.634e+04   1.17e+05     -0.398      0.691   -2.75e+05    1.83e+05
furnishingstatus_unfurnished    -4.112e+05   1.26e+05     -3.258      0.001   -6.59e+05   -1.63e+05
==============================================================================
Omnibus:                       97.909   Durbin-Watson:                   1.209
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              258.281
Skew:                           0.895   Prob(JB):                     8.22e-57
Kurtosis:                       5.859   Cond. No.                     3.49e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.49e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

当我们把显著区间设定为0.05时，以上结果的P值可以看出，模型认为以下因素对房屋价格没有显著性影响：卧室数、是否为简装房。此外，常数（表示线性方程的截距）的P值也很大，说明也没有显著影响。

可以把这些变量移除后，再次建立线性回归模型。

In [24]:
x=x.drop(['const','bedrooms','furnishingstatus_semi-furnished'],axis=1)

In [25]:
model=sm.OLS(y,x).fit()

In [26]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  price   R-squared (uncentered):                   0.957
Model:                            OLS   Adj. R-squared (uncentered):              0.956
Method:                 Least Squares   F-statistic:                              1088.
Date:                Wed, 20 Nov 2024   Prob (F-statistic):                        0.00
Time:                        15:47:04   Log-Likelihood:                         -8333.5
No. Observations:                 545   AIC:                                  1.669e+04
Df Residuals:                     534   BIC:                                  1.674e+04
Df Model:                          11                                                  
Covariance Type:            nonrobust                                                  
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
area                           257.5854     22.577     11.409      0.000     213.234     301.936
bathrooms                     1.071e+06   9.27e+04     11.553      0.000    8.89e+05    1.25e+06
stories                       5.084e+05   5.71e+04      8.898      0.000    3.96e+05    6.21e+05
parking                       2.793e+05   5.83e+04      4.794      0.000    1.65e+05    3.94e+05
mainroad_yes                  4.672e+05   1.27e+05      3.679      0.000    2.18e+05    7.17e+05
guestroom_yes                 2.851e+05   1.31e+05      2.172      0.030    2.72e+04    5.43e+05
basement_yes                  4.016e+05   1.07e+05      3.765      0.000    1.92e+05    6.11e+05
hotwaterheating_yes           8.668e+05   2.23e+05      3.884      0.000    4.28e+05    1.31e+06
airconditioning_yes           8.543e+05   1.07e+05      7.952      0.000    6.43e+05    1.07e+06
prefarea_yes                  6.443e+05   1.15e+05      5.594      0.000    4.18e+05    8.71e+05
furnishingstatus_unfurnished -3.493e+05   9.49e+04     -3.679      0.000   -5.36e+05   -1.63e+05
==============================================================================
Omnibus:                       94.840   Durbin-Watson:                   1.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              251.889
Skew:                           0.865   Prob(JB):                     2.01e-55
Kurtosis:                       5.845   Cond. No.                     2.74e+04
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large, 2.74e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

可以看到，当我们把P值较大的自变量从线性回归模型中移除后，R方的值从0.682增长到了0.957，提高了模型对现有数据的拟合度。

根据各个自变量在线性回归方程中的系数来看，模型预测以下因素的增加（或存在）会显著增加房屋价格：房屋面积、厕所数、楼层数、车库容量、位于主路、有客房、有地下室、有热水器、有空调、位于城市首选社区。

线性回归模型预测以下因素的增加（或存在）会显著降低房屋价格：房屋未经装修，为毛坯房。

In [38]:
# 要预测房价的房屋的信息：
# 面积为6500平方英尺，有4个卧室、2个厕所，总共2层，不位于主路，无客人房，带地下室，有热水器，没有空调，车位数为2，位于城市首选社区，简装修

In [39]:
price_to_predict=pd.DataFrame({'area':[6500],'bedrooms':[4],'bathrooms': [2], 
                                 'stories': [2], 'mainroad': ['no'], 'guestroom': ['no'],
                                 'basement': ['yes'], 'hotwaterheating': ['yes'],
                                 'airconditioning': ['no'], 'parking': 2, 'prefarea': ['yes'],
                                 'furnishingstatus': ['semi-furnished']})
price_to_predict

area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  6500         4          2        2       no        no      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0             yes              no        2      yes   semi-furnished

我们需要把分类变量的类型转换为Category，并且通过categories参数，让程序知道所有可能的分类值。这样做的原因是，预测数据包含的分类可能不全。我们需要确保引入虚拟变量的时候，不会漏掉某个或某些分类。

In [40]:
price_to_predict['mainroad'] = pd.Categorical(price_to_predict['mainroad'], categories=['no', 'yes'])
price_to_predict['guestroom'] = pd.Categorical(price_to_predict['guestroom'], categories=['no', 'yes'])
price_to_predict['basement'] = pd.Categorical(price_to_predict['basement'], categories=['no', 'yes'])
price_to_predict['hotwaterheating'] = pd.Categorical(price_to_predict['hotwaterheating'], categories=['no', 'yes'])
price_to_predict['airconditioning'] = pd.Categorical(price_to_predict['airconditioning'], categories=['no', 'yes'])
price_to_predict['prefarea'] = pd.Categorical(price_to_predict['prefarea'], categories=['no', 'yes'])
price_to_predict['furnishingstatus'] = pd.Categorical(price_to_predict['furnishingstatus'], categories=['furnished', 'semi-furnished', 'unfurnished'])

对分类变量引入虚拟变量。

In [41]:
price_to_predict=pd.get_dummies(price_to_predict,
                                columns=['mainroad', 'guestroom',
                                           'basement', 'hotwaterheating',
                                           'airconditioning','prefarea', 
                                           'furnishingstatus'],dtype=int,drop_first=True)
price_to_predict.head()

area  bedrooms  bathrooms  stories  parking  mainroad_yes  guestroom_yes  \
0  6500         4          2        2        2             0              0   

   basement_yes  hotwaterheating_yes  airconditioning_yes  prefarea_yes  \
0             1                    1                    0             1   

   furnishingstatus_semi-furnished  furnishingstatus_unfurnished  
0                                1                             0

In [45]:
price_to_predict=price_to_predict.drop(['bedrooms','furnishingstatus_semi-furnished'],axis=1)

接下来就可以调用线性回归模型的predict方法，获得预测价格。

In [46]:
predicted_value = model.predict(price_to_predict)
predicted_value

0    7.303754e+06
dtype: float64

线性回归模型预测的价格为：7071927。